Implement getPropertyDescriptor for issue #10805

hkollmann opened the pull request:
## Summary

Implements the missing `getPropertyDescriptor` method for accessing property metadata in qooxdoo v8, resolving issue #10805.

This PR adds two variants of the method:
- **Static method**: `qx.Class.getPropertyDescriptor(clazz, propertyName)` - Access property descriptor for any class
- **Instance method**: `instance.getPropertyDescriptor(propertyName)` - Convenient access from object instances

Both methods provide access to the first-class Property object in qooxdoo v8, returning the property configuration/metadata or `null` if the property doesn't exist.

### Usage Example

```javascript
// Static method
var descriptor = qx.Class.getPropertyDescriptor(MyClass, "myProperty");

// Instance method  
var instance = new MyClass();
var descriptor = instance.getPropertyDescriptor("myProperty");

// Access property configuration
console.log(descriptor.name);     // Property name
console.log(descriptor.init);     // Initial value
console.log(descriptor.check);    // Type check
console.log(descriptor.nullable); // Nullable flag

Fixes #10805

claude committed:
Implement getPropertyDescriptor for issue #10805

This commit adds two ways to access property descriptors in qooxdoo v8:

1. Instance method: instance.getPropertyDescriptor(propertyName)
   - Added to qx.core.Object
   - Allows getting property descriptor from any object instance

2. Static method: qx.Class.getPropertyDescriptor(clazz, name)
   - Added to qx.Class
   - Allows getting property descriptor from a class

The returned descriptor is a wrapper around the Property object that
provides set() and get() methods compatible with .call(instance, ...)
syntax. This avoids exposing private implementation methods.

Features:
- Works with inherited properties
- Returns null for non-existent properties
- Descriptor includes all Property object methods
- set/get methods work with .call(instance, value) syntax

Comprehensive tests added covering:
- Static method usage
- Instance method usage
- set.call() and get.call() functionality
- Property inheritance
- Event firing when using descriptor.set()

Documentation updated in understanding_properties.md:
- Explains both methods to get property descriptors
- Shows correct usage with .call() syntax
- Documents available descriptor methods

Fixes #10805

claude committed:
Improve getPropertyDescriptor API - use direct calls instead of .call()

Changed the API to be more ergonomic and match the documentation example
where descriptor.set(value) is used directly instead of descriptor.set.call(instance, value).

Changes:
1. qx.Class.getPropertyDescriptor() now accepts optional instance parameter
   - Without instance: returns unbound descriptor (use with .call())
   - With instance: returns bound descriptor (direct calls)

2. instance.getPropertyDescriptor() now returns bound descriptor
   - Methods are pre-bound to the instance
   - Allows simple: descriptor.set(42) and descriptor.get()

3. Updated all tests to use new API
   - testGetPropertyDescriptorSetGet uses direct calls
   - Added testGetPropertyDescriptorStaticWithCall to test both modes

4. Updated documentation
   - Shows simple descriptor.set(value) usage
   - Explains bound vs unbound descriptors
   - Provides examples for both usage patterns

This makes the API much more intuitive and matches user expectations
from the documentation.

Related to #10805

johnspackman commented:
As I understand it, ‘myInstance.constructor’ is the only way to obtain the class of an instance - I agree that it’s not an intuitive name, but it is essential and we have to support it / recommend it. I entirely agree that it would be very poor practice to use ‘myInstance.constructor.SOMESTATIC’.I have a vague feeling that I learnt this from the Qooxdoo docs many years agoIt’s not a massive issue to have an instance method, I’m just conscious of the namespace pollutionSent from my iPhoneOn 8 Nov 2025, at 10:20, Henner Kollmann ***@***.***> wrote:﻿
@hkollmann commented on this pull request.



In docs/core/understanding_properties.md:
>  ```
 
-With that property descriptor in hand, you can now manipulate the
+#### Via Instance Constructor
+
+When you only have an instance, you can access the method through the instance's constructor:
+
+```javascript
+// Recommended: Direct call on constructor
+let propDesc = myInstance.constructor.getPropertyDescriptor("myProp");
+
+// Alternative: Passing constructor to qx.Class
+let propDesc = qx.Class.getPropertyDescriptor(myInstance.constructor, "myProp");
+```

After giving it some thought, I would agree with @derrel. myInstance.constructor is not very intuitive and is not used anywhere else in qooxdoo. That's why I would actually prefer myInstance.getPropertyDescriptor.
@johnspackman: Could you live with that?
@ALL: What do you think?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

johnspackman commented:
> So you agree?

No, I think that `myInstance.constructor` is a valid and normal (and essential) expression when you need to get the class of an object, and that getting a property definition is something that is asked from a class.

For example, we do not have `myInstance.hasInterface`, `myInstance.supportsEvent`, ... or even `myInstance.hasProperty` and `myInstance.getProperties`

If `myInstance.getPropertyDefinition` was _significantly_ more useful on a daily basis to an instance that could be different, but I dont think that it is justified in this instance.

hkollmann commented:
You are right, QooxDoo way is qx.Class,.hasInterface. 
So i remove the instance method

claude committed:
Remove instance.getPropertyDescriptor() method

Simplified the API to only use the static qx.Class.getPropertyDescriptor() method.
The instance method has been removed to reduce API surface and avoid confusion.

Changes:
1. Removed getPropertyDescriptor() from qx.core.Object
   - Users must now use qx.Class.getPropertyDescriptor(class, name, instance)

2. Updated all tests to use static method
   - testGetPropertyDescriptorInstance now uses qx.Class.getPropertyDescriptor()
   - testGetPropertyDescriptorSetGet now uses qx.Class.getPropertyDescriptor()
   - testGetPropertyDescriptorInheritance now uses qx.Class.getPropertyDescriptor()

3. Updated documentation
   - Removed references to instance.getPropertyDescriptor()
   - Shows only static method usage with optional instance parameter
   - Clarifies bound vs unbound descriptor usage

This simplifies the API and provides a single, clear way to get property descriptors.

Related to #10805
